### PR TITLE
Fix #5465 by catching no-image exception

### LINKED ISSUE #5465: Building container fails after removing old image.
I found this issue while working on a fix for issue #4337.

By running the following sequence of commands on [this example repository](https://github.com/twelve17/dc-test):

```
git checkout tags/test2_run1
docker-compose build
docker rmi --force dctest_service1
docker-compose build
docker-compose up
```

Compose fails because the old image is never properly rebuilt from the old container:

```
WARNING: The Docker Engine you're using is running in swarm mode.

Compose does not use swarm mode to deploy services to multiple nodes in a swarm. All containers will be scheduled on the current node.

To deploy your application across the swarm, use `docker stack deploy`.

Recreating dctest_service1_1 ... error

ERROR: for dctest_service1_1  no such image: sha256:bb41503c2e723478308998afc7df0886fac0811621b70d61b28ca36418a41bb7: No such image: sha256:bb41503c2e723478308998afc7df0886fac0811621b70d61b28ca36418a41bb7

ERROR: for service1  no such image: sha256:bb41503c2e723478308998afc7df0886fac0811621b70d61b28ca36418a41bb7: No such image: sha256:bb41503c2e723478308998afc7df0886fac0811621b70d61b28ca36418a41bb7
ERROR: Encountered errors while bringing up the project.
```

### PR DESCRIPTION
**Pull request by Dahca**
Fix #5465 by catching a no-image exception in get_container_data_volumes. Added a case to the unit test to check that an empty existing image doesn't result in an exception.

Closed old pull request #5466 because I did a rebase and it showed over 300 commits to merge, which I thought was messy.

### CODE REVIEW

**Comment by shin-** [comment]
@Dahca Thanks! The messy rebase was probably from when I re-aligned master with the release branch :confounded: sorry.

Do you mind squashing these commits together? Other than that I think we're good to go.

**Comment by Dahca** [comment]
Hey @shin-, not a problem!

Is this squashed enough? I was trying to squash the whole thing, but when I do it always has a merge conflict when I do the rebase from upstream.

**Comment by Dahca** [comment]
Looks like jenkins timed out -- is there a way to re-run the check?

File: compose/cli/main.py [hunk 1]
```diff
@@ -979,21 +979,23 @@ def up(self, options):
...
         with up_shutdown_context(self.project, service_names, timeout, detached):
-            try:
-                to_attach = self.project.up(
+            def up(rebuild):
+                return self.project.up(
                     service_names=service_names,
                     start_deps=start_deps,
                     strategy=convergence_strategy_from_opts(options),
-                    do_build=build_action_from_opts(options),
+                    do_build=build_action_from_opts(options) if not rebuild else BuildAction.force,
```
Comment (Dahca) [inline compose/cli/main.py]: This actually doesn't fix the problem. When the build is forced, it still checks the images and throws the NoImageFound exception.

**Comment by Dahca** [comment]
@shin- The bug is still present even with BuildAction.force.

File: compose/cli/main.py (hunk 1 above)
Comment (shin-) [reply compose/cli/main.py]: Ah, that sucks :(

Can we add a test for it? I know there's a unit test there but it didn't catch anything with my changes, so we probably need something more comprehensive. Once we have a good test we can figure out what the best approach to fixing it is.

Comment (Dahca) [reply compose/cli/main.py]: Alright, I'll work on that.

File: tests/unit/service_test.py [hunk 2]
```diff
@@ -926,7 +927,7 @@ def test_get_container_data_volumes(self):
         volumes, _ = get_container_data_volumes(container, options, ['/dev/tmpfs'], [])
         assert sorted(volumes) == sorted(expected)
 
-    def test_get_container_data_volumes_image_is_none(self):
+    def test_get_container_data_volumes_image_does_not_exist(self):
```
Comment (Dahca) [inline tests/unit/service_test.py]: @shin- Here's the updated test, which currently fails.

Comment (shin-) [reply tests/unit/service_test.py]: Mmmh ok, I'll take a look over the next couple of days. Thank you!

**Comment by shin-** [comment]
Merged in #5583 - thank you for your help and patience!

**Pull request closed without merging**